components/net/local_directory_listing.rs

```diff
@@ -0,0 +1,430 @@
+        // forward-slash to the end of the URL (at least internally,
+        // though the URL bar does not reflect this) and loads the
+        // linked directories and files correctly when clicked.
+        if let Ok(dir_url) = Url::from_directory_path(url.path()) {
+            ServoUrl::from_url(dir_url)
+        } else {
+            return Response::network_error(NetworkError::Internal(format!(
+                "Unable to parse local directory path {}",
+                url
+            )));
+        }
+    } else {
+        url
+    };
+    let mut response = Response::new(url, ResourceFetchTiming::new(request.timing_type()));
+    response.headers.typed_insert(ContentType::html());
+    let (path, has_parent, items) = summarise_directory(path_buf);
+    let page_text = build_html_directory_listing(path, has_parent, items);
+    let bytes: Vec<u8> = page_text.into_bytes();
+    *response.body.lock().unwrap() = ResponseBody::Done(bytes);
+    response
+}
+
+/// Returns information about the given local directory.
+///
+/// # Arguments
+///
+/// `path_buf` - a `PathBuf` which points to a local directory path.
+///
+/// # Returns
+///
+/// A tuple containing three members (path, has_parent, items):
+///
+/// * `path` - contains the full path to the local directory, or an error message if the path
+/// cannot be successfully rendered as a `String`.
+/// * `has_parent` - will be `true` if the local directory has a parent directory, or `false` if
+/// the local directory is the top-level directory or cannot reach its parent for any reason.
+/// * `items` - will contain a map of directory items in alphabetical order by name, or an error
+/// message if the local directory path could not be iterated for any reason.
+pub fn summarise_directory(
+    path_buf: PathBuf,
+) -> (
+    Result<String, &'static str>,
+    bool,
+    Result<BTreeMap<OsString, DirectoryItem>, &'static str>,
+) {
+    let path = path_buf
+        .to_str()
+        .map(str::to_string)
+        .ok_or("Invalid directory path.");
+    let has_parent = path_buf.parent().is_some();
+    let items = if let Ok(entries) = std::fs::read_dir(path_buf) {
+        Ok(gather_directory_items(entries))
+    } else {
+        Err("Unable to iterate directory contents.")
+    };
+    (path, has_parent, items)
+}
+
+fn gather_directory_items(entries: std::fs::ReadDir) -> BTreeMap<OsString, DirectoryItem> {
+    let map: BTreeMap<OsString, DirectoryItem> = entries
+        .into_iter()
+        .filter_map(|e| e.ok())
+        .flat_map(|e| create_name_item_mapping(e))
+        .collect();
+    map
+}
+
+fn create_name_item_mapping(entry: DirEntry) -> Option<(OsString, DirectoryItem)> {
+    if let Ok(meta) = entry.metadata() {
+        let os_name = entry.file_name();
+        let entry_name = os_name.to_str().map(str::to_string);
+        create_directory_item(entry_name, meta).map(|i| (os_name, i))
+    } else {
+        None
+    }
```

Comment:
```suggestion
    if let Ok(metadata) = entry.metadata() else {
        return;
    }
    let os_name = entry.file_name();
    let entry_name = os_name.to_str().map(str::to_string);
    create_directory_item(entry_name, metadata).map(|i| (os_name, i))
```

Reply:
I can't get the suggested version to compile, so I've reworked my code to get rid of the if/else and use a match pattern instead. That puts the `None=>None` outcome at the first line of the body, in the same way as your suggested code does.

Reply:
What is the error that you get when compiling?

Reply:
Syntax errors, return type errors, and variable scoping errors:

```
error: this `if` expression is missing a block after the condition
   --> components/net/local_directory_listing.rs:113:5
    |
113 |     if let Ok(metadata) = entry.metadata() else {
    |     ^^
    |
help: add a block here
   --> components/net/local_directory_listing.rs:113:43
    |
113 |     if let Ok(metadata) = entry.metadata() else {
    |                                           ^
help: remove the `if` if you meant to write a `let...else` statement
    |
113 -     if let Ok(metadata) = entry.metadata() else {
113 +     let Ok(metadata) = entry.metadata() else {
    |

error[E0425]: cannot find value `metadata` in this scope
   --> components/net/local_directory_listing.rs:118:39
    |
118 |     create_directory_item(entry_name, metadata).map(|i| (os_name, i))
    |                                       ^^^^^^^^ not found in this scope
    |
help: consider importing one of these items
    |
5   + use core::ptr::metadata;
    |
5   + use std::fs::metadata;
    |
5   + use std::ptr::metadata;
    |
5   + use tokio::fs::metadata;
    |

error[E0069]: `return;` in a function whose return type is not `()`
   --> components/net/local_directory_listing.rs:114:9
    |
112 | fn create_name_item_mapping(entry: DirEntry) -> Option<(OsString, DirectoryItem)> {
    |                                                 --------------------------------- expected `Option<(OsString, DirectoryItem)>` because of this return type
113 |     if let Ok(metadata) = entry.metadata() else {
114 |         return;
    |         ^^^^^^ return type is not `()`

Some errors have detailed explanations: E0069, E0425.
For more information about an error, try `rustc --explain E0069`.
error: could not compile `net` (lib) due to 3 previous errors
```

Reply:
I've replaced the whole function with this:

```
fn create_name_item_mapping(entry: DirEntry) -> Option<(OsString, DirectoryItem)> {
    match entry.metadata() {
        Err(_) => None,
        Ok(metadata) => {
            let os_name = entry.file_name();
            let entry_name = os_name.to_str().map(str::to_string);
            create_directory_item(entry_name, metadata).map(|i| (os_name, i))
        },
    }
}
```

If this is no good, let me know what else it needs.